### ISSUE TITLE
`write_json` failed with struct containing `Eigen::Matrix3d`

### ISSUE THREAD

**Issue by CHN-beta** [stephenberry/glaze#1457]
```c++
# include <glaze/glaze.hpp>
# include <Eigen/Dense>
# include <glaze/ext/eigen.hpp>

struct test_struct
{
  Eigen::Matrix3d d = Eigen::Matrix3d::Identity();
} test;

int main()
{
  auto result = glz::write_json(test.d).value(); // <- this works
  // auto result = glz::write_json(test).value(); // <- this does not work, with the following errors
}
```

Serialize `test_struct` instead of its member leading to the following error:

```
[build] In file included from /home/chn/repo/nixos/packages/biu/test/glaze.cpp:3:
[build] In file included from /nix/store/192041k004z7is2v7c29mgjq715m6w74-eigen-3.4.0/include/eigen3/Eigen/Dense:1:
[build] In file included from /nix/store/192041k004z7is2v7c29mgjq715m6w74-eigen-3.4.0/include/eigen3/Eigen/Core:272:
[build] In file included from /nix/store/192041k004z7is2v7c29mgjq715m6w74-eigen-3.4.0/include/eigen3/Eigen/src/Core/DenseBase.h:660:
[build] In file included from /nix/store/192041k004z7is2v7c29mgjq715m6w74-eigen-3.4.0/include/eigen3/Eigen/src/Core/../plugins/IndexedViewMethods.h:210:
[build] /nix/store/192041k004z7is2v7c29mgjq715m6w74-eigen-3.4.0/include/eigen3/Eigen/src/Core/../plugins/IndexedViewMethods.h:157:35: error: static assertion failed due to requirement 'Matrix<double, 3, 3, 0, 3, 3>::IsVectorAtCompileTime': YOU_TRIED_CALLING_A_VECTOR_METHOD_ON_A_MATRIX
[build]   157 |   EIGEN_STATIC_ASSERT_VECTOR_ONLY(Derived)
[build]       |   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~
[build] /nix/store/192041k004z7is2v7c29mgjq715m6w74-eigen-3.4.0/include/eigen3/Eigen/src/Core/util/StaticAssert.h:143:23: note: expanded from macro 'EIGEN_STATIC_ASSERT_VECTOR_ONLY'
[build]   143 |   EIGEN_STATIC_ASSERT(TYPE::IsVectorAtCompileTime, \
[build]       |   ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[build]   144 |                       YOU_TRIED_CALLING_A_VECTOR_METHOD_ON_A_MATRIX)
[build]       |                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[build] /nix/store/192041k004z7is2v7c29mgjq715m6w74-eigen-3.4.0/include/eigen3/Eigen/src/Core/util/StaticAssert.h:33:54: note: expanded from macro 'EIGEN_STATIC_ASSERT'
[build]    33 |     #define EIGEN_STATIC_ASSERT(X,MSG) static_assert(X,#MSG);
[build]       |                                                      ^
[build] /nix/store/605hgp0v6rf9945vmvz2pizz069h5bvc-gcc-13.3.0/include/c++/13.3.0/bits/invoke.h:61:14: note: in instantiation of function template specialization 'Eigen::DenseBase<Eigen::Matrix<double, 3, 3>>::operator()<aaa>' requested here
[build]    61 |     { return std::forward<_Fn>(__f)(std::forward<_Args>(__args)...); }
[build]       |              ^
[build] /nix/store/0s596n6znalhk88bdnq8l2xribrx4jib-glaze-4.0.1/include/glaze/json/write.hpp:37:56: note: in instantiation of function template specialization 'glz::detail::to<10, aaa>::op<opts{10, 1, 0, 1, 1, 1, 0, 0, 32, 3, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0}, aaa &, std::basic_string<char> &, glz::context &, unsigned long &>' requested here
[build]    37 |             to<JSON, std::remove_cvref_t<T>>::template op<Opts>(std::forward<T>(value), std::forward<Ctx>(ctx),
[build]       |                                                        ^
[build] /nix/store/0s596n6znalhk88bdnq8l2xribrx4jib-glaze-4.0.1/include/glaze/core/write.hpp:24:44: note: in instantiation of function template specialization 'glz::detail::write<10>::op<opts{10, 1, 0, 1, 1, 1, 0, 0, 32, 3, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0}, aaa &, glz::context &, std::basic_string<char> &, unsigned long &>' requested here
[build]    24 |       detail::write<Opts.format>::template op<Opts>(std::forward<T>(value), ctx, buffer, ix);
[build]       |                                            ^
[build] /nix/store/0s596n6znalhk88bdnq8l2xribrx4jib-glaze-4.0.1/include/glaze/core/write.hpp:77:23: note: in instantiation of function template specialization 'glz::write<opts{10, 1, 0, 1, 1, 1, 0, 0, 32, 3, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0}, aaa &, std::basic_string<char>, glz::context &>' requested here
[build]    77 |       const auto ec = write<Opts>(std::forward<T>(value), buffer, ctx);
[build]       |                       ^
[build] /nix/store/0s596n6znalhk88bdnq8l2xribrx4jib-glaze-4.0.1/include/glaze/json/write.hpp:1787:14: note: in instantiation of function template specialization 'glz::write<opts{10, 1, 0, 1, 1, 1, 0, 0, 32, 3, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0}, aaa &>' requested here
[build]  1787 |       return write<opts{}>(std::forward<T>(value));
[build]       |              ^
[build] /home/chn/repo/nixos/packages/biu/test/glaze.cpp:17:22: note: in instantiation of function template specialization 'glz::write_json<aaa &>' requested here
[build]    17 |   auto result = glz::write_json(bbb).value();
[build]       |                      ^
[build] In file included from /home/chn/repo/nixos/packages/biu/test/glaze.cpp:3:
[build] In file included from /nix/store/192041k004z7is2v7c29mgjq715m6w74-eigen-3.4.0/include/eigen3/Eigen/Dense:1:
[build] In file included from /nix/store/192041k004z7is2v7c29mgjq715m6w74-eigen-3.4.0/include/eigen3/Eigen/Core:162:
[build] /nix/store/192041k004z7is2v7c29mgjq715m6w74-eigen-3.4.0/include/eigen3/Eigen/src/Core/util/Meta.h:479:51: error: no member named 'size' in 'aaa'
[build]   479 | EIGEN_CONSTEXPR Index size(const T& x) { return x.size(); }
[build]       |                                                 ~ ^
[build] /nix/store/192041k004z7is2v7c29mgjq715m6w74-eigen-3.4.0/include/eigen3/Eigen/src/Core/IndexedView.h:125:41: note: in instantiation of function template specialization 'Eigen::internal::size<aaa>' requested here
[build]   125 |   Index rows() const { return internal::size(m_rowIndices); }
[build]       |                                         ^
[build] /nix/store/192041k004z7is2v7c29mgjq715m6w74-eigen-3.4.0/include/eigen3/Eigen/src/Core/EigenBase.h:60:63: note: in instantiation of member function 'Eigen::IndexedView<Eigen::Matrix<double, 3, 3>, aaa, Eigen::internal::SingleRange>::rows' requested here
[build]    60 |   inline Index rows() const EIGEN_NOEXCEPT { return derived().rows(); }
[build]       |                                                               ^
[build] /nix/store/0s596n6znalhk88bdnq8l2xribrx4jib-glaze-4.0.1/include/glaze/json/write.hpp:1570:51: note: in instantiation of function template specialization 'glz::detail::to<10, Eigen::IndexedView<Eigen::Matrix<double, 3, 3>, aaa, Eigen::internal::SingleRange>>::op<opts{10, 1, 0, 1, 1, 1, 0, 0, 32, 3, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0}, std::basic_string<char> &, Eigen::IndexedView<Eigen::Matrix<double, 3, 3>, aaa, Eigen::internal::SingleRange>, glz::context &, unsigned long &>' requested here
[build]  1570 |                         to<JSON, val_t>::template op<check_opts>(get_member(value, get<I>(t)), ctx, b, ix);
[build]       |                                                   ^
[build] /nix/store/0s596n6znalhk88bdnq8l2xribrx4jib-glaze-4.0.1/include/glaze/json/write.hpp:37:56: note: in instantiation of function template specialization 'glz::detail::to<10, aaa>::op<opts{10, 1, 0, 1, 1, 1, 0, 0, 32, 3, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0}, aaa &, std::basic_string<char> &, glz::context &, unsigned long &>' requested here
[build]    37 |             to<JSON, std::remove_cvref_t<T>>::template op<Opts>(std::forward<T>(value), std::forward<Ctx>(ctx),
[build]       |                                                        ^
[build] /nix/store/0s596n6znalhk88bdnq8l2xribrx4jib-glaze-4.0.1/include/glaze/core/write.hpp:24:44: note: in instantiation of function template specialization 'glz::detail::write<10>::op<opts{10, 1, 0, 1, 1, 1, 0, 0, 32, 3, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0}, aaa &, glz::context &, std::basic_string<char> &, unsigned long &>' requested here
[build]    24 |       detail::write<Opts.format>::template op<Opts>(std::forward<T>(value), ctx, buffer, ix);
[build]       |                                            ^
[build] /nix/store/0s596n6znalhk88bdnq8l2xribrx4jib-glaze-4.0.1/include/glaze/core/write.hpp:77:23: note: in instantiation of function template specialization 'glz::write<opts{10, 1, 0, 1, 1, 1, 0, 0, 32, 3, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0}, aaa &, std::basic_string<char>, glz::context &>' requested here
[build]    77 |       const auto ec = write<Opts>(std::forward<T>(value), buffer, ctx);
[build]       |                       ^
[build] /nix/store/0s596n6znalhk88bdnq8l2xribrx4jib-glaze-4.0.1/include/glaze/json/write.hpp:1787:14: note: in instantiation of function template specialization 'glz::write<opts{10, 1, 0, 1, 1, 1, 0, 0, 32, 3, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0}, aaa &>' requested here
[build]  1787 |       return write<opts{}>(std::forward<T>(value));
[build]       |              ^
[build] /home/chn/repo/nixos/packages/biu/test/glaze.cpp:17:22: note: in instantiation of function template specialization 'glz::write_json<aaa &>' requested here
[build]    17 |   auto result = glz::write_json(bbb).value();
[build]       |                      ^
[build] In file included from /home/chn/repo/nixos/packages/biu/test/glaze.cpp:3:
[build] In file included from /nix/store/192041k004z7is2v7c29mgjq715m6w74-eigen-3.4.0/include/eigen3/Eigen/Dense:1:
[build] In file included from /nix/store/192041k004z7is2v7c29mgjq715m6w74-eigen-3.4.0/include/eigen3/Eigen/Core:311:
[build] /nix/store/192041k004z7is2v7c29mgjq715m6w74-eigen-3.4.0/include/eigen3/Eigen/src/Core/IndexedView.h:207:32: error: type 'const aaa' does not provide a subscript operator
[build]   207 |     return m_argImpl.coeffRef( m_xpr.rowIndices()[row], m_xpr.colIndices()[col]);
[build]       |                                ^~~~~~~~~~~~~~~~~~ ~~~
[build] /nix/store/192041k004z7is2v7c29mgjq715m6w74-eigen-3.4.0/include/eigen3/Eigen/src/Core/DenseCoeffsBase.h:394:54: note: in instantiation of member function 'Eigen::internal::unary_evaluator<Eigen::IndexedView<Eigen::Matrix<double, 3, 3>, aaa, Eigen::internal::SingleRange>>::coeffRef' requested here
[build]   394 |       return internal::evaluator<Derived>(derived()).coeffRef(index);
[build]       |                                                      ^
[build] /nix/store/192041k004z7is2v7c29mgjq715m6w74-eigen-3.4.0/include/eigen3/Eigen/src/Core/DenseCoeffsBase.h:428:14: note: in instantiation of member function 'Eigen::DenseCoeffsBase<Eigen::IndexedView<Eigen::Matrix<double, 3, 3>, aaa, Eigen::internal::SingleRange>, 1>::coeffRef' requested here
[build]   428 |       return coeffRef(index);
[build]       |              ^
[build] /nix/store/192041k004z7is2v7c29mgjq715m6w74-eigen-3.4.0/include/eigen3/Eigen/src/Core/StlIterators.h:299:50: note: in instantiation of member function 'Eigen::DenseCoeffsBase<Eigen::IndexedView<Eigen::Matrix<double, 3, 3>, aaa, Eigen::internal::SingleRange>, 1>::operator()' requested here
[build]   299 |   reference operator*()         const { return   (*mp_xpr)(m_index);   }
[build]       |                                                  ^
[build] /nix/store/0s596n6znalhk88bdnq8l2xribrx4jib-glaze-4.0.1/include/glaze/json/write.hpp:741:76: note: in instantiation of member function 'Eigen::internal::generic_randaccess_stl_iterator<Eigen::IndexedView<Eigen::Matrix<double, 3, 3>, aaa, Eigen::internal::SingleRange>>::operator*' requested here
[build]   741 |                   to<JSON, val_t>::template op<write_unchecked_on<Opts>()>(*it, ctx, b, ix);
[build]       |                                                                            ^
[build] /nix/store/0s596n6znalhk88bdnq8l2xribrx4jib-glaze-4.0.1/include/glaze/json/write.hpp:1570:51: note: in instantiation of function template specialization 'glz::detail::to<10, Eigen::IndexedView<Eigen::Matrix<double, 3, 3>, aaa, Eigen::internal::SingleRange>>::op<opts{10, 1, 0, 1, 1, 1, 0, 0, 32, 3, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0}, std::basic_string<char> &, Eigen::IndexedView<Eigen::Matrix<double, 3, 3>, aaa, Eigen::internal::SingleRange>, glz::context &, unsigned long &>' requested here
[build]  1570 |                         to<JSON, val_t>::template op<check_opts>(get_member(value, get<I>(t)), ctx, b, ix);
[build]       |                                                   ^
[build] /nix/store/0s596n6znalhk88bdnq8l2xribrx4jib-glaze-4.0.1/include/glaze/json/write.hpp:37:56: note: in instantiation of function template specialization 'glz::detail::to<10, aaa>::op<opts{10, 1, 0, 1, 1, 1, 0, 0, 32, 3, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0}, aaa &, std::basic_string<char> &, glz::context &, unsigned long &>' requested here
[build]    37 |             to<JSON, std::remove_cvref_t<T>>::template op<Opts>(std::forward<T>(value), std::forward<Ctx>(ctx),
[build]       |                                                        ^
[build] /nix/store/0s596n6znalhk88bdnq8l2xribrx4jib-glaze-4.0.1/include/glaze/core/write.hpp:24:44: note: in instantiation of function template specialization 'glz::detail::write<10>::op<opts{10, 1, 0, 1, 1, 1, 0, 0, 32, 3, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0}, aaa &, glz::context &, std::basic_string<char> &, unsigned long &>' requested here
[build]    24 |       detail::write<Opts.format>::template op<Opts>(std::forward<T>(value), ctx, buffer, ix);
[build]       |                                            ^
[build] /nix/store/0s596n6znalhk88bdnq8l2xribrx4jib-glaze-4.0.1/include/glaze/core/write.hpp:77:23: note: in instantiation of function template specialization 'glz::write<opts{10, 1, 0, 1, 1, 1, 0, 0, 32, 3, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0}, aaa &, std::basic_string<char>, glz::context &>' requested here
[build]    77 |       const auto ec = write<Opts>(std::forward<T>(value), buffer, ctx);
[build]       |                       ^
[build] /nix/store/0s596n6znalhk88bdnq8l2xribrx4jib-glaze-4.0.1/include/glaze/json/write.hpp:1787:14: note: in instantiation of function template specialization 'glz::write<opts{10, 1, 0, 1, 1, 1, 0, 0, 32, 3, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0}, aaa &>' requested here
[build]  1787 |       return write<opts{}>(std::forward<T>(value));
[build]       |              ^
[build] /home/chn/repo/nixos/packages/biu/test/glaze.cpp:17:22: note: in instantiation of function template specialization 'glz::write_json<aaa &>' requested here
[build]    17 |   auto result = glz::write_json(bbb).value();
[build]       |                      ^
[build] 3 errors generated.
```

Tested with clang 18, with glaze 4.0.1 and Eigen 3.4.0.

**Comment by stephenberry** [comment]
Thanks for reporting this. It turns out it is due to bad template meta-programming in Eigen. The `operator()()` is telling the compiler that `Eigen::Matrix3d` can invoke its owning parent class, which is how Glaze sets up lambdas to change serialization layout. However, `Eigen::Matrix3d` actually can't invoke the parent class, the developers just add a `static_assert` inside the `operator()()` call. So, meta-programming will tell the compiler that `Eigen::Matrix3d` can invoke all kinds of types, but then will always cause the compilation to fail. The static assertion should really have been an `enable_if` on the operator (a concept in modern C++).

Now I'm looking to see if there is a work-around.

**Comment by CHN-beta** [comment]
> I merged a fix (#1458) that rejects matrix types from the invocable check, so your test case now passes.

Great job! Thank you for your kindly response!
